### PR TITLE
[CDAP-17466] Added supported for optionally running an Authentication service in k8s.

### PR DESCRIPTION
**Pull request by chtyim**
This one replaces #59 with test fixes.

### CODE REVIEW

File: api/v1alpha1/cdapmaster_types.go [hunk 1]
```diff
@@ -146,7 +163,7 @@ type MessagingSpec struct {
 
 // MetadataSpec defines the specification for the Metadata service
 type MetadataSpec struct {
-	CDAPStatefulServiceSpec `json:",inline"`
+	CDAPScalableServiceSpec `json:",inline"`
```
Comment (wyzhang) [inline api/v1alpha1/cdapmaster_types.go:166]: Thanks for fixing the mistake here.

File: api/v1alpha1/cdapmaster_types.go [hunk 2]
```diff
@@ -98,6 +107,14 @@ type CDAPServiceSpec struct {
...
 	// Env is a list of environment variables for the master service container.
 	Env []corev1.EnvVar `json:"env,omitempty"`
+	// ConfigMapVolumes defines a map from ConfigMap names to volume mount path for this service
+	// Key is the configmap object name. Value is the mount path.
+	// This adds ConfigMap data to the directory specified by the volume mount path.
+	ConfigMapVolumes map[string]string `json:"configMapVolumes,omitempty"`
+	// SecretVolumes defines a map from Secret names to volume mount path for this service
+	// Key is the secret object name. Value is the mount path.
+	// This adds Secret data to the directory specified by the volume mount path.
+	SecretVolumes map[string]string `json:"secretVolumes,omitempty"`
```
Comment (wyzhang) [inline api/v1alpha1/cdapmaster_types.go:117]: Are these used for overriding those settings in CDAPMasterSpec ? If so, mind updating the comment to state that explicitly?

If I recall correctly, ConfigMapVolumes defined in CDAPMasterSpec will mount config map to all statefulsets and deployments.

File: controllers/deployment.go [hunk 3]
```diff
@@ -94,6 +95,7 @@ func buildDeploymentPlanSpec(master *v1alpha1.CDAPMaster, labels map[string]stri
 	spec := newDeploymentPlanSpec()
 	// Build statefulsets
 	for k, v := range serviceGroups.stateful {
+
```
Comment (wyzhang) [inline controllers/deployment.go]: Remove the redundant new line?

File: templates/cdap-deployment.yaml [hunk 4]
```diff
@@ -106,6 +106,10 @@ spec:
             - name: cdap-cm-vol-{{$k}}
               mountPath: {{$v}}
             {{end}}
+            {{range $k,$v := $.Base.SecretVolumes}}
+            - name: cdap-secret-vol-{{$k}}
```
Comment (wyzhang) [inline templates/cdap-deployment.yaml]: Consider shortening it to cdap-se-vol-

Recall I ran into name string length limit issue when calling it cdap-configmap-vol-

File: templates/cdap-sts.yaml [hunk 5]
```diff
@@ -146,6 +146,10 @@ spec:
             - name: cdap-cm-vol-{{$k}}
               mountPath: {{$v}}
             {{end}}
+            {{range $k,$v := $.Base.SecretVolumes}}
+            - name: cdap-secret-vol-{{$k}}
```
Comment (wyzhang) [inline templates/cdap-sts.yaml]: Same here.

File: controllers/spec.go [hunk 6]
```diff
@@ -148,7 +150,16 @@ func newBaseSpec(master *v1alpha1.CDAPMaster, name string, labels map[string]str
 	s.CConf = cconf
 	s.HConf = hconf
 	s.SysAppConf = sysappconf
-	s.ConfigMapVolumes = master.Spec.ConfigMapVolumes
+
+	s.ConfigMapVolumes = make(map[string]string)
+	for k, v := range master.Spec.ConfigMapVolumes {
+		s.ConfigMapVolumes[k] = v
+	}
```
Comment (wyzhang) [inline controllers/spec.go]: Agree that better to deep copy maps. Curious if shadow copy caused any sigsegv during your test, thus making this change?

Comment (wyzhang) [reply controllers/spec.go]: Mind helping with converting the label copy to deep copy as well while you are here. Or feel free to leave as is that I can clean it up across the entire codebase in a separate change

**Review by wyzhang — APPROVED** [review]

File: api/v1alpha1/cdapmaster_types.go (hunk 2 above)
Comment (chtyim) [reply api/v1alpha1/cdapmaster_types.go:117]: It is not for overriding, but appending/merging.

File: controllers/spec.go (hunk 6 above)
Comment (chtyim) [reply controllers/spec.go]: Done.

I haven't tried shallow copy, but I think it is better to keep the `master` spec not mutated.

File: templates/cdap-deployment.yaml (hunk 4 above)
Comment (chtyim) [reply templates/cdap-deployment.yaml]: The name limit is 253 characters. Did we run issue before?

File: controllers/deployment.go (hunk 3 above)
Comment (chtyim) [reply controllers/deployment.go]: done

**Pull request merged**